File: Design Patterns/01 Strategy Pattern/C#/01_strategy_pattern/01_strategy_pattern/ShoppingCart.cs
using System;
using System.Collections.Generic;
using System.Linq;
using _01_strategy_pattern.interfaces;

namespace _01_strategy_pattern
{
    internal sealed class ShoppingCart
    {
        internal List<Tuple<IItem, IPaymentMethod>> _items;

        public ShoppingCart()
        {
            this._items = new List<Tuple<IItem, IPaymentMethod>>();
        }

        public void AddItem(IItem item, IPaymentMethod paymentMethod)
        {
            this._items.Add(new Tuple<IItem, IPaymentMethod>(item, paymentMethod));
            Console.WriteLine($"Added Item '{item.Name}' with PaymentMethod {paymentMethod.getPaymentMethodName()} into ShoppingCart.");
        }

        public Tuple<IItem, IPaymentMethod> getItem(Guid itemIdentifier)
        {
            return _items.FirstOrDefault(i => i.Item1.Identifier == itemIdentifier);
        }

        public void removeItem(Tuple<IItem, IPaymentMethod> itemTuple)
        {
            if (itemTuple != null)
            {
                _items.Remove(itemTuple);
                Console.WriteLine($"Removed Item '{itemTuple.Item1.Name}' from ShoppingCart.");
            }
            else
            {
                Console.WriteLine($"Item 'UNKNOWN not found in ShoppingCart.");
            }
        }

        public void Checkout()
        {
            foreach (var item in _items)
            {
                Console.WriteLine($"Item {item.Item1.Name} will be payed using {item.Item2.getPaymentMethodName()}");
            }
        }   
    }
}

File: Design Patterns/01 Strategy Pattern/python/01_strategy_pattern/ShoppingCart.py
from typing import List, Tuple
from interfaces import BasePaymentMethod, BaseItem


class ShoppingCart:
    items: List[Tuple[BaseItem, BasePaymentMethod]]

    def __init__(self):
        self.items = []

    def additem(self, item: BaseItem, paymentmethod: BasePaymentMethod):
        self.items.append((item, paymentmethod))
        print(f'Added Item "{item.name}" with PaymentMethod "{paymentmethod.getpaymentmethodname()}" into ShoppingCart.')

    def removeitem(self, itemtuple: Tuple[BaseItem, BasePaymentMethod]):
        if itemtuple in self.items:
            self.items.remove(itemtuple)
            print(f'Removed Item "{itemtuple[0].name}" from ShoppingCart.')
        else:
            print(f'Item "UNKNOWN" not found in ShoppingCart.')

    def checkout(self):
        for item, payment_method in self.items:
            print(f'Item {item.name} will be paid using {payment_method.getpaymentmethodname()}')